cli/ssh.go

```diff
@@ -1085,3 +1096,49 @@ func getUsageAppName(usageApp string) codersdk.UsageAppName {
 
 	return codersdk.UsageAppNameSSH
 }
+
+type timeoutCloser struct {
+	target      io.Closer
+	closeCalled chan struct{}
+
+	// for testing
+	clock quartz.Clock
+}
+
+func newTimeoutCloser(
+	ctx context.Context, logger slog.Logger, timeout time.Duration, target io.Closer, clock quartz.Clock,
+) *timeoutCloser {
+	b := &timeoutCloser{
+		target:      target,
+		closeCalled: make(chan struct{}),
+		clock:       clock,
+	}
+	go b.waitForCtxOrClose(ctx, logger, timeout)
+	return b
+}
+
+func (t *timeoutCloser) waitForCtxOrClose(ctx context.Context, logger slog.Logger, timeout time.Duration) {
+	select {
+	case <-t.closeCalled:
+		return
+	case <-ctx.Done():
+	}
+	tmr := t.clock.NewTimer(timeout, "timeoutCloser", "waitForCtxOrClose")
+	defer tmr.Stop()
+	select {
+	case <-t.closeCalled:
+		return
+	case <-tmr.C:
+		logger.Warn(ctx, "timed out waiting for graceful shutdown")
+		err := t.target.Close()
+		if err != nil {
+			logger.Debug(ctx, "error closing target", slog.Error(err))
+		}
+	}
+}
+
+// Close should only be called at most once, e.g. in the closerStack
+func (t *timeoutCloser) Close() error {
+	close(t.closeCalled)
+	return t.target.Close()
```

Comment:
This implementation has a theoretical shortcoming that I think we should try to avoid considering the issue we're fixing.

If the operation that hangs is actually the call to `target.Close()`, and we reach this point of execution, `waitForCtxOrClose` will exit on `t.closeCalled` and this method may hang forever. I don't think it will happen, but I'd still feel better if we returned from this `Close` after timeout.

---

That said, I think the implementation could be simplified and made more robust by moving the timeout to closerStack.

Motivation: If bottom closure fails to propagate up the stack, we will still hang while closing the stack.

`(*closerStack).Close` would instead look ~like this (partial implementation for demo):

```go
go func() {
	defer close(done)
	for i := len(c.closers) - 1; i >= 0; i-- {
		cwn := c.closers[i]
		cErr := cwn.closer.Close()
		c.logger.Debug(context.Background(),
			"closed item from stack", slog.F("name", cwn.name), slog.Error(cErr))
	}
}()
select {
case <-ctx.Done():
case <-done:
}
```

I'm aware this will ultimately not call `conn.Close()` and clean up resources, but at this point, does it really even matter? We're about to exit the process anyway.

Alternatively it could be a two-pronged approach where first you try to close within the time limit, if that fails simultaneously call all close methods with a new timeout. Or give each close method their own timeout, depending on how clean we want the shutdown to be.

Reply:
> I'm aware this will ultimately not call conn.Close() and clean up resources, but at this point, does it really even matter? We're about to exit the process anyway.

I was going to say it would be nice to still clean up the agentconn/tailnet `Conn`, since we send a `Disconnect` message to the `Coordinator` to indicate that we're intentionally shutting down. But, thinking about it a bit more, the known scenario where this bites us is if we've lost connectivity anyway.

I do like the simplicity of just bailing out of the `closerStack`, and I agree it's good enough for our purposes and more robust.

Reply:
I tried the simple approach of just bailing out, and unfortunately, that doesn't solve the high level issue, because the SSH function also has a `WaitGroup` waiting for stuff to exit, so if closing truly hangs, so does waiting for the `WaitGroup`.

It might make sense to add functionality to Serpent to handle graceful shutdown at a higher layer, so we can call `os.Exit()` if things hang.  I'm guessing there are a few other CLI commands that could use a more robust solution.

Reply:
I did take the advice of integrating it all into the closerStack, so we will ensure that each item gets called exactly once, and we will eventually time out of closing the stack.